### PR TITLE
feat: show github issue's labels

### LINKED ISSUE #674: Show labels from github
For issues included from github, it might be nice to also include the labels.

### PR DESCRIPTION
**Pull request by himanish-star**
fixes #674 
Preview of changes:
image

### CODE REVIEW

File: src/core/issues-notes.js [hunk 1]
```diff
@@ -118,7 +118,37 @@ function handleIssues($ins, ghIssues, conf) {
...
+              const color = label.color.substring(1);
+              const rgb = parseInt(color, 16);
+              const red = (rgb >> 16) & 0xff;
+              const green = (rgb >>  8) & 0xff;
+              const blue = (rgb >>  0) & 0xff;
+              const illumination = 0.2126 * red + 0.7152 * green + 0.0722 * blue;
+              if (illumination > 140) {
+                labelSpan.style.color = '#000000'
+              } else {
+                labelSpan.style.color = '#ffffff'
```
Comment (himanish-star) [inline src/core/issues-notes.js]: This part of the code does this:
If the background color is too dark, the text color is set to white
else is set to black

File: tests/spec/core/issues-notes-spec.js [hunk 2]
```diff
@@ -126,6 +126,30 @@ describe("Core — Issues and Notes", function () {
...
+    };
+    const doc = await makeRSDoc(ops);
+    const issueDiv = doc.getElementById('issue-1540');
+    expect(issueDiv).toBeTruthy();
+    const label0 = doc.getElementById('1540-label-0');
+    expect(label0).toBeTruthy();
+    expect(label0.innerText).toEqual("refactor");
+    const label1 = doc.getElementById('1540-label-1');
+    expect(label1).toBeTruthy();
+    expect(label1.innerText).toEqual("bug");
```
Comment (himanish-star) [inline tests/spec/core/issues-notes-spec.js]: I guess there is no need to check for the colors of the label and yes should I also add the test for the issue with number 404?

File: src/core/issues-notes.js (hunk 1 above)
Comment (marcoscaceres) [reply src/core/issues-notes.js]: Ah, that’s cute. Is this something you came up with?

Comment (marcoscaceres) [reply src/core/issues-notes.js]: Also, we should not set style directly. Set a css class instead, which sets the color to “white” or “black”.

File: src/core/issues-notes.js [hunk 3]
```diff
@@ -118,7 +118,37 @@ function handleIssues($ins, ghIssues, conf) {
...
+          var labelListSpan = document.createElement('span');
+          Array.from(ghIssue.labels)
+            .forEach((label, index) => {
+              var labelSpan = document.createElement('span');
+              labelSpan.classList.add("respec-gh-label");
+              labelSpan.style.backgroundColor = `#${label.color}`;
+              labelSpan.style.textTransform = 'none';
+              labelSpan.id = `${dataNum}-label-${index}`;
+              labelSpan.innerText = label.name;
+              const color = label.color.substring(1);
```
Comment (marcoscaceres) [inline src/core/issues-notes.js]: All the color code should move to it’s own “deriveTextColorClass()” that returns “white” or “black”.

File: tests/data/issues/1540 [hunk 4]
```diff
@@ -1,5 +1,15 @@
 {
   "number": 1540,
   "title": "A mock open issue for testing",
-  "state": "open"
+  "state": "open",
+  "labels": [
```
Comment (marcoscaceres) [inline tests/data/issues/1540]: 👌

File: src/core/issues-notes.js [hunk 5]
```diff
@@ -118,7 +118,37 @@ function handleIssues($ins, ghIssues, conf) {
           }
         }
         $tit.find("span").text(text);
-        if (report.title) {
+        if (report.title && githubAPI) {
+          var labelListSpan = document.createElement('span');
```
Comment (marcoscaceres) [inline src/core/issues-notes.js]: If it’s a list, then we might want a `ul`, but I don’t think we want a list at all

File: src/core/issues-notes.js [hunk 6]
```diff
@@ -118,7 +118,37 @@ function handleIssues($ins, ghIssues, conf) {
...
+                labelSpan.style.color = '#000000'
+              } else {
+                labelSpan.style.color = '#ffffff'
+              }
+              labelListSpan.appendChild(labelSpan);
+            });
+          $tit.append(
+            $(
+              "<span style='text-transform: none'>: " + report.title + labelListSpan.innerHTML
+              +"</span>"
```
Comment (marcoscaceres) [inline src/core/issues-notes.js]: The text-transform css should be in the  stylesheet. As a rule, never inline CSS.

File: src/core/issues-notes.js [hunk 7]
```diff
@@ -118,7 +118,37 @@ function handleIssues($ins, ghIssues, conf) {
           }
         }
         $tit.find("span").text(text);
-        if (report.title) {
+        if (report.title && githubAPI) {
+          var labelListSpan = document.createElement('span');
+          Array.from(ghIssue.labels)
+            .forEach((label, index) => {
```
Comment (marcoscaceres) [inline src/core/issues-notes.js]: This “forEach” creates things (labels), so it’s a ` .map(createLabel)`.  And those labels get reduce()’d into the outter container.

File: src/core/issues-notes.js (hunk 5 above)
Comment (marcoscaceres) [reply src/core/issues-notes.js]: If you want an arbitrary container, use a DocumentFragment.

File: tests/spec/core/issues-notes-spec.js (hunk 2 above)
Comment (marcoscaceres) [reply tests/spec/core/issues-notes-spec.js]: Yep

File: src/core/issues-notes.js [hunk 8]
```diff
@@ -118,7 +118,37 @@ function handleIssues($ins, ghIssues, conf) {
...
+              var labelSpan = document.createElement('span');
```
Comment (marcoscaceres) [inline src/core/issues-notes.js]: Labels should be `a` elements, as the value of labels is that they group issues together. Clicking on a label should take you to github, so you can see other issues with the same label.

File: src/core/issues-notes.js [hunk 9]
```diff
@@ -118,7 +118,37 @@ function handleIssues($ins, ghIssues, conf) {
...
+              const blue = (rgb >>  0) & 0xff;
+              const illumination = 0.2126 * red + 0.7152 * green + 0.0722 * blue;
+              if (illumination > 140) {
+                labelSpan.style.color = '#000000'
+              } else {
+                labelSpan.style.color = '#ffffff'
+              }
+              labelListSpan.appendChild(labelSpan);
+            });
+          $tit.append(
```
Comment (marcoscaceres) [inline src/core/issues-notes.js]: let's pull the actual HTML element out of `$tit`, so to avoid relying on more jQuery.

File: src/core/issues-notes.js [hunk 10]
```diff
@@ -118,7 +118,37 @@ function handleIssues($ins, ghIssues, conf) {
...
+            $(
+              "<span style='text-transform: none'>: " + report.title + labelListSpan.innerHTML
```
Comment (marcoscaceres) [inline src/core/issues-notes.js]: Note here that we went to all the trouble of creating HTML objects, so it's a shame to have to change them to a string to add them to the container.

File: src/core/issues-notes.js [hunk 11]
```diff
@@ -118,7 +118,37 @@ function handleIssues($ins, ghIssues, conf) {
...
-        if (report.title) {
+        if (report.title && githubAPI) {
+          var labelListSpan = document.createElement('span');
+          Array.from(ghIssue.labels)
+            .forEach((label, index) => {
+              var labelSpan = document.createElement('span');
+              labelSpan.classList.add("respec-gh-label");
+              labelSpan.style.backgroundColor = `#${label.color}`;
+              labelSpan.style.textTransform = 'none';
+              labelSpan.id = `${dataNum}-label-${index}`;
```
Comment (marcoscaceres) [inline src/core/issues-notes.js]: An issue can appear in the document more than once, so this would lead to duplicate IDs. That would make the document invalid as per the W3C Validator. 

We probably don't want to give these things IDs anyway (I can't think of any use case for having an id).

File: src/core/issues-notes.js [hunk 12]
```diff
@@ -118,7 +118,37 @@ function handleIssues($ins, ghIssues, conf) {
...
         $tit.find("span").text(text);
-        if (report.title) {
+        if (report.title && githubAPI) {
+          var labelListSpan = document.createElement('span');
+          Array.from(ghIssue.labels)
+            .forEach((label, index) => {
+              var labelSpan = document.createElement('span');
+              labelSpan.classList.add("respec-gh-label");
+              labelSpan.style.backgroundColor = `#${label.color}`;
+              labelSpan.style.textTransform = 'none';
```
Comment (marcoscaceres) [inline src/core/issues-notes.js]: This should be in the style sheet instead.

File: src/core/issues-notes.js [hunk 13]
```diff
@@ -118,7 +118,37 @@ function handleIssues($ins, ghIssues, conf) {
...
+              labelSpan.id = `${dataNum}-label-${index}`;
+              labelSpan.innerText = label.name;
+              const color = label.color.substring(1);
+              const rgb = parseInt(color, 16);
```
Comment (marcoscaceres) [inline src/core/issues-notes.js]: What happens if this is NaN?

File: src/core/issues-notes.js [hunk 14]
```diff
@@ -118,7 +118,37 @@ function handleIssues($ins, ghIssues, conf) {
...
+              const red = (rgb >> 16) & 0xff;
+              const green = (rgb >>  8) & 0xff;
+              const blue = (rgb >>  0) & 0xff;
+              const illumination = 0.2126 * red + 0.7152 * green + 0.0722 * blue;
```
Comment (marcoscaceres) [inline src/core/issues-notes.js]: Please add a comment here that this works out the grayscale value.

File: tests/data/issues/1540 [hunk 15]
```diff
@@ -1,5 +1,15 @@
...
+      {
+        "name": "refactor",
+        "color": "47f441"
+      },
+      {
+        "name": "bug",
+        "color": "f4425c"
+      }
```
Comment (marcoscaceres) [inline tests/data/issues/1540]: For completeness, we should add a `"color": ""` test case, and a `"color": "this is not a color"` test case.

File: src/core/issues-notes.js [hunk 16]
```diff
@@ -163,7 +193,7 @@ async function fetchIssuesFromGithub({ githubAPI }) {
       if (!response.ok) {
         switch (response.status) {
           case 404:
-            throw new Error("Couldn't find issue on Github. Check if it exist?");
+            throw new Error("Couldn't find issue on Github. Check if it exists?");
```
Comment (marcoscaceres) [inline src/core/issues-notes.js]: Maybe send this fix as a different PR? We can get this one merged in quickly.

File: src/core/issues-notes.js (hunk 7 above)
Comment (himanish-star) [reply src/core/issues-notes.js]: > This “forEach” creates things (labels), so it’s a .map(createLabel). And those labels get reduce()’d into the outer container.

Thanks for this, looks great to implement.

File: src/core/issues-notes.js (hunk 3 above)
Comment (marcoscaceres) [reply src/core/issues-notes.js]: Actually, it should return “light” or “dark”.

Comment (himanish-star) [reply src/core/issues-notes.js]: Yep got it. I will update this in a while.

File: src/core/issues-notes.js (hunk 11 above)
Comment (himanish-star) [reply src/core/issues-notes.js]: I was actually using the ids in testing :stuck_out_tongue:

File: tests/data/issues/1540 (hunk 15 above)
Comment (himanish-star) [reply tests/data/issues/1540]: Couldn't understand this :disappointed:

Comment (marcoscaceres) [reply tests/data/issues/1540]: A good option might be to have a fallback color for when things go wrong. E.g., white text, black background.

Comment (himanish-star) [reply tests/data/issues/1540]: Ah fine :smiley:

Comment (himanish-star) [reply tests/data/issues/1540]: I have pushed a part of the changes now. I am adding this feature now

File: src/core/issues-notes.js [hunk 17]
```diff
@@ -118,7 +118,17 @@ function handleIssues($ins, ghIssues, conf) {
           }
         }
         $tit.find("span").text(text);
-        if (report.title) {
+        if (report.title && githubAPI) {
+          const labelsGroup = Array.from(ghIssue.labels)
+            .map(createLabels)
+            .reduce(groupLabels);
```
Comment (himanish-star) [inline src/core/issues-notes.js]: Is this what you wanted, I mean in the way I have used map and reduce

File: src/core/issues-notes.js [hunk 18]
```diff
@@ -146,6 +156,40 @@ function handleIssues($ins, ghIssues, conf) {
...
+
+function createLabels(label) {
+  const textColorClass = deriveTextColorClass(label.color);
+  var labelElement = document.createElement("a");
+  labelElement.classList.add("respec-gh-label");
+  labelElement.style.backgroundColor = `#${label.color}`;
+  labelElement.style.textTransform = "none";
+  labelElement.href = label.url;
+  labelElement.innerText = label.name;
+  labelElement.classList.add(`respec-label-${textColorClass}`);
```
Comment (himanish-star) [inline src/core/issues-notes.js]: The CSS of this class has no role to play as it get's overwritten. How do I fix this? The UI looks very unpleasing without the CSS of this class.

File: src/core/issues-notes.js [hunk 19]
```diff
@@ -118,7 +118,20 @@ function handleIssues($ins, ghIssues, conf) {
           }
         }
         $tit.find("span").text(text);
-        if (report.title) {
+        if (report.title && githubAPI) {
+          const labelsGroup = Array.from(ghIssue.labels)
+            .map(createLabels)
+            .reduce((frag, labelElem) => {
```
Comment (marcoscaceres) [inline src/core/issues-notes.js]: When in doubt: 

screenshot 2018-03-19 13 33 25

Where the sandwich is a `DocumentFragment` ;)

File: src/core/issues-notes.js [hunk 20]
```diff
@@ -146,6 +159,33 @@ function handleIssues($ins, ghIssues, conf) {
...
+  const red = (rgb >> 16) & 0xff;
+  const green = (rgb >> 8) & 0xff;
+  const blue = (rgb >> 0) & 0xff;
+  const illumination = 0.2126 * red + 0.7152 * green + 0.0722 * blue;
+  return illumination > 140 ? "light" : "dark";
+}
+
+function createLabels(label) {
+  const hexColor = label.color.substring(1);
+  var textColorClass = "dark";
```
Comment (marcoscaceres) [inline src/core/issues-notes.js]: let

File: src/core/issues-notes.js [hunk 21]
```diff
@@ -146,6 +159,33 @@ function handleIssues($ins, ghIssues, conf) {
...
+  var labelElement = document.createElement("a");
```
Comment (marcoscaceres) [inline src/core/issues-notes.js]: const

File: src/core/issues-notes.js [hunk 22]
```diff
@@ -146,6 +159,33 @@ function handleIssues($ins, ghIssues, conf) {
...
+  if (!isNaN(parseInt(hexColor, 16))) {
```
Comment (marcoscaceres) [inline src/core/issues-notes.js]: parseInt is being done twice: here and in `deriveTextColorClass`. You should only do it in  `deriveTextColorClass()` and throw if NaN. Then here, you should deal with that exception.

File: src/core/issues-notes.js (hunk 19 above)
Comment (himanish-star) [reply src/core/issues-notes.js]: Oh

File: src/core/issues-notes.js (hunk 22 above)
Comment (himanish-star) [reply src/core/issues-notes.js]: Yep got it

**Comment by himanish-star** [comment]
I have made the changes. Now we got to focus on the UI changes because based on this I'll be writing the tests which are now failing due to the proper CSS changes not reflecting
Currently the UI looks like this :stuck_out_tongue: 

image

File: src/core/issues-notes.js [hunk 23]
```diff
@@ -118,7 +118,20 @@ function handleIssues($ins, ghIssues, conf) {
...
+        if (report.title && githubAPI) {
+          const labelsGroup = Array.from(ghIssue.labels)
+            .map(createLabels)
+            .reduce((frag, labelElem) => {
+              frag.appendChild(labelElem);
+              return frag;
+            }, document.createDocumentFragment());
+          $tit.append(
+            $(
+              "<span style='text-transform: none'>: " + report.title + "</span>"
```
Comment (himanish-star) [inline src/core/issues-notes.js]: This I had copied from the existing code, so in that this was inline styling. So, if you want me to make a class for this then I would have to make a class for this. If yes, suggest me a name for this class.

File: src/core/issues-notes.js [hunk 24]
```diff
@@ -146,6 +159,36 @@ function handleIssues($ins, ghIssues, conf) {
...
+    if (isNaN(rgb)) {
+      throw NaN;
+    }
+    const red = (rgb >> 16) & 0xff;
+    const green = (rgb >> 8) & 0xff;
+    const blue = (rgb >> 0) & 0xff;
+    const illumination = 0.2126 * red + 0.7152 * green + 0.0722 * blue;
+    return illumination > 140 ? "light" : "dark";
+  } catch (err) {
+    console.error(err);
```
Comment (himanish-star) [inline src/core/issues-notes.js]: Maybe we can improve the text content of `console.error()`

Comment (marcoscaceres) [reply src/core/issues-notes.js]: Remember that console.error() is not for end-users, it's just for ReSpec maintainers so we can debug any issues. Generally, we only need `err` and the stack trace to figure out what happened.

File: src/core/issues-notes.js [hunk 25]
```diff
@@ -146,6 +159,36 @@ function handleIssues($ins, ghIssues, conf) {
   }
 }
 
+//derives the text-color-class based on the illumination score of the labelColor
+function deriveTextColorClass(hexColor) {
+  try {
+    const rgb = parseInt(hexColor, 16);
+    if (isNaN(rgb)) {
+      throw NaN;
```
Comment (marcoscaceres) [inline src/core/issues-notes.js]: ProTip: never throw anything apart from an actual Error or something that inherits from Error (e.g., `TypeError`). Here we want a type error... 

Also, making the whole body of the function a try/catch is not a good idea. We don't want to mask programming errors, with input errors.  

Thus, this should just be:

```JS
function deriveTextColorClass(hexColor) {
  const rgb = parseInt(hexColor, 16);
  if (isNaN(rgb)) {
    throw new TypeError(`Invalid hex color: ${hexColor}`);
  }
  const red = (rgb >> 16) & 0xff;
  const green = (rgb >> 8) & 0xff;
  const blue = (rgb >> 0) & 0xff;
  const illumination = 0.2126 * red + 0.7152 * green + 0.0722 * blue;
  return illumination > 140 ? "light" : "dark";
}
```

File: src/core/issues-notes.js (hunk 23 above)
Comment (marcoscaceres) [reply src/core/issues-notes.js]: if this is a `.issue`, then a `.issue .title, .note .title` rule would be good.

File: src/core/issues-notes.js (hunk 25 above)
Comment (marcoscaceres) [reply src/core/issues-notes.js]: to be clear, you always want `Error`, because otherwise you don't get a stack trace.

File: src/core/issues-notes.js [hunk 26]
```diff
@@ -146,6 +159,36 @@ function handleIssues($ins, ghIssues, conf) {
...
+    const blue = (rgb >> 0) & 0xff;
+    const illumination = 0.2126 * red + 0.7152 * green + 0.0722 * blue;
+    return illumination > 140 ? "light" : "dark";
+  } catch (err) {
+    console.error(err);
+    return "dark";
+  }
+}
+
+function createLabels(label) {
```
Comment (marcoscaceres) [inline src/core/issues-notes.js]: How about instead:

```JS
import "deps/hyperhtml"; // import this at the top

function createLabels(label) {
  const { color, url, name } = label;
  let textColorClass = "dark"; 
  try {
    textColorClass = deriveTextColorClass(color);
  } catch(err) {
     console.error(err);
  }
  const cssClasses = `respec-gh-label respec-label-${textColorClass}`;
  return hyperHTML`<a
    class="${cssClasses}"
    style="background-color: #${color}"
    href="${url}">${name}</a>`;
}
```

File: tests/data/issues/1540 [hunk 27]
```diff
@@ -1,5 +1,23 @@
 {
   "number": 1540,
   "title": "A mock open issue for testing",
-  "state": "open"
+  "state": "open",
+  "labels": [
+      {
+        "name": "refactor",
+        "color": "47f441"
```
Comment (marcoscaceres) [inline tests/data/issues/1540]: Some of these should include a `url:` prop, as it's expected with `createLabels()`.

Comment (himanish-star) [reply tests/data/issues/1540]: Yep sure

File: tests/spec/core/issues-notes-spec.js [hunk 28]
```diff
@@ -126,6 +126,38 @@ describe("Core — Issues and Notes", function () {
...
+        <div class='issue' data-number='1540'>issue is open on github</div>
+        <div class='issue' id='this-is-404' data-number='404'>this is 404</div>
+        <section id='issue-summary'></section>
+        `,
+    };
+    const doc = await makeRSDoc(ops);
+    const issueDiv = doc.getElementById("issue-1540");
+    expect(issueDiv).toBeTruthy();
+    const labels = doc.getElementsByClassName("respec-gh-label");
+    expect(labels[0].innerText).toEqual("refactor");
```
Comment (marcoscaceres) [inline tests/spec/core/issues-notes-spec.js]: nit: it's nice to name each label, so we can then more easily know what which is which. So:

```JS
const [refactorLabel, bugLabel, blankLabel] = doc.querySelectorAll(
  ".respec-gh-label"
);
expect(bugLabel.innerText).toEqual("bug");

```

File: tests/spec/core/issues-notes-spec.js [hunk 29]
```diff
@@ -126,6 +126,38 @@ describe("Core — Issues and Notes", function () {
...
+    expect(labels[1].innerText).toEqual("bug");
```
Comment (marcoscaceres) [inline tests/spec/core/issues-notes-spec.js]: Don't forget to check the `href` attribute for each.

File: js/ui/ui.css [hunk 30]
```diff
@@ -93,6 +93,18 @@
   font-size: 0.8em;
 }
 
+.respec-gh-label {
```
Comment (marcoscaceres) [inline js/ui/ui.css]: Just realized that this CSS is in the wrong place.  It should be moved to:

```
js/core/css/issues-notes.css
```

The "js/ui/ui.css" is only for the ReSpec pill, and is removed from the document on save.

Comment (marcoscaceres) [reply js/ui/ui.css]: To get these to show up:

```CSS
.issue a.respec-gh-label {
  padding: 5px;
  margin: 0 2px 0 2px;
  font-size: 10px;
  text-transform: none;
  text-decoration: none;
  font-weight: bold;
  border-radius: 4px;
  position: relative;
  bottom: 2px;
}

.issue a.respec-label-dark {
  color: #fff;
  background-color: #000;
}

.issue a.respec-label-light {
  color: #000;
  background-color: #fff;
}

```

File: src/core/issues-notes.js [hunk 31]
```diff
@@ -118,7 +119,20 @@ function handleIssues($ins, ghIssues, conf) {
           }
         }
         $tit.find("span").text(text);
-        if (report.title) {
+        if (report.title && githubAPI) {
+          const labelsGroup = Array.from(ghIssue.labels)
+            .map(createLabels)
```
Comment (marcoscaceres) [inline src/core/issues-notes.js]: nit: rename "createLabels" to "createLabel"

File: src/core/issues-notes.js [hunk 32]
```diff
@@ -146,6 +160,35 @@ function handleIssues($ins, ghIssues, conf) {
...
+    throw new TypeError(`Invalid hex color: ${hexColor}`);
+  }
+  const red = (rgb >> 16) & 0xff;
+  const green = (rgb >> 8) & 0xff;
+  const blue = (rgb >> 0) & 0xff;
+  const illumination = 0.2126 * red + 0.7152 * green + 0.0722 * blue;
+  return illumination > 140 ? "light" : "dark";
+}
+
+function createLabels(label) {
```
Comment (marcoscaceres) [inline src/core/issues-notes.js]: This should be "createLabel" instead... as it only creates one at a time.

File: src/core/issues-notes.js [hunk 33]
```diff
@@ -146,6 +160,35 @@ function handleIssues($ins, ghIssues, conf) {
...
+  const { color, url, name } = label;
```
Comment (marcoscaceres) [inline src/core/issues-notes.js]: Sorry, I misunderstood what `url` was for... that's not the URL we want :) we need to construct a different one above.

File: src/core/issues-notes.js [hunk 34]
```diff
@@ -118,7 +119,20 @@ function handleIssues($ins, ghIssues, conf) {
           }
         }
         $tit.find("span").text(text);
-        if (report.title) {
+        if (report.title && githubAPI) {
```
Comment (marcoscaceres) [inline src/core/issues-notes.js]: here we actually want:
```
 if (ghIssue && report.title && githubAPI) {
```

Comment (himanish-star) [reply src/core/issues-notes.js]: Yeah sure

File: src/core/issues-notes.js (hunk 33 above)
Comment (himanish-star) [reply src/core/issues-notes.js]: Yep

File: src/core/issues-notes.js (hunk 32 above)
Comment (himanish-star) [reply src/core/issues-notes.js]: Ok

File: src/core/issues-notes.js (hunk 31 above)
Comment (himanish-star) [reply src/core/issues-notes.js]: yep

File: js/ui/ui.css (hunk 30 above)
Comment (himanish-star) [reply js/ui/ui.css]: Yeah sure

File: tests/data/issues/1540 (hunk 27 above)
Comment (himanish-star) [reply tests/data/issues/1540]: Now I guess there is no need of the `url:` prop, as we are generating our own urls

File: src/core/issues-notes.js [hunk 35]
```diff
@@ -146,6 +168,35 @@ function handleIssues($ins, ghIssues, conf) {
   }
 }
 
+//derives the text-color-class based on the illumination score of the labelColor
+function deriveTextColorClass(hexColor) {
+  const rgb = parseInt(hexColor, 16);
+  if (isNaN(rgb)) {
+    throw new TypeError(`Invalid hex color: ${hexColor}`);
```
Comment (himanish-star) [inline src/core/issues-notes.js]: Thanks for explaining this

File: src/core/issues-notes.js [hunk 36]
```diff
@@ -118,7 +119,28 @@ function handleIssues($ins, ghIssues, conf) {
           }
         }
         $tit.find("span").text(text);
-        if (report.title) {
+        if (ghIssue && report.title && githubAPI) {
+          const labelsGroup = Array.from(ghIssue.labels)
+            .map(label => {
+              const issuesURL = new URL('issues/', conf.github + '/');
```
Comment (himanish-star) [inline src/core/issues-notes.js]: added this '/' otherwise the urls were like https://github.com/w3c/ instead of https://github.com/w3c/respec

File: tests/data/issues/1540 (hunk 27 above)
Comment (marcoscaceres) [reply tests/data/issues/1540]: Correct!

File: src/core/issues-notes.js (hunk 35 above)
Comment (marcoscaceres) [reply src/core/issues-notes.js]: No problem. Trust me, you'll be thanking me for years :) Keep an eye out for that from other developers. It can make code super difficult to debug, specially async code.

**Comment by marcoscaceres** [comment]
Whoops, I accidentally merged without squashing :(

**Comment by marcoscaceres** [comment]
+ respec@20.2.0 

Squashed on develop... because that's how we roll.

**Comment by marcoscaceres** [comment]
We hit some bugs with this, so going to revert it.

**Comment by himanish-star** [comment]
Oh
